File: app/src/main/java/agersant/polaris/CollectionItem.java
package agersant.polaris;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.io.Serializable;
import java.lang.reflect.Type;

public class CollectionItem implements Cloneable, Serializable {

    private String path;
    private String artist;
    private String title;
    private String artwork;
    private String album;
    private String albumArtist;
    private String composer;
    private String lyricist;
    private String genre;
    private String label;
    private int trackNumber;
    private int discNumber;
    private int duration;
    private int year;
    @SuppressWarnings("WeakerAccess")
    boolean isDirectory;

    private CollectionItem() {
    }

    private static String getNameFromPath(String path) {
        String[] chunks = path.split("[/\\\\]");
        return chunks[chunks.length - 1];
    }

    public static CollectionItem directory(String path) {
        CollectionItem item = new CollectionItem();
        item.isDirectory = true;
        item.path = path;
        return item;
    }

    void parseFields(JsonObject fields) {
        path = getOptionalString(fields, "path");
        artist = getOptionalString(fields, "artist");
        title = getOptionalString(fields, "title");
        artwork = getOptionalString(fields, "artwork");
        album = getOptionalString(fields, "album");
        albumArtist = getOptionalString(fields, "album_artist");
        composer = getOptionalString(fields, "composer");
        lyricist = getOptionalString(fields, "lyricist");
        genre = getOptionalString(fields, "genre");
        label = getOptionalString(fields, "label");
        trackNumber = getOptionalInt(fields, "track_number");
        discNumber = getOptionalInt(fields, "disc_number");
        duration = getOptionalInt(fields, "duration");
        year = getOptionalInt(fields, "year");
    }

    private String getOptionalString(JsonObject fields, String key) {
        if (!fields.has(key)) {
            return null;
        }
        JsonElement element = fields.get(key);
        if (element.isJsonNull()) {
            return null;
        }
        return element.getAsString();
    }

    private int getOptionalInt(JsonObject fields, @SuppressWarnings("SameParameterValue") String key) {
        if (!fields.has(key)) {
            return -1;
        }
        JsonElement element = fields.get(key);
        if (element.isJsonNull()) {
            return -1;
        }
        return element.getAsInt();
    }

    public String getName() {
        return getNameFromPath(path);
    }

    @Override
    public CollectionItem clone() throws CloneNotSupportedException {
        return (CollectionItem) super.clone();
    }

    public String getPath() {
        return path;
    }

    public String getTitle() {
        return title;
    }

    public String getArtist() {
        return artist;
    }

    public String getArtwork() {
        return artwork;
    }

    public String getAlbum() {
        return album;
    }

    public String getAlbumArtist() {
        return albumArtist;
    }

    public String getComposer() {
        return composer;
    }

    public String getLyricist() {
        return lyricist;
    }

    public String getGenre() {
        return genre;
    }

    public String getLabel() {
        return label;
    }

    public int getTrackNumber() {
        return trackNumber;
    }

    public int getDiscNumber() {
        return discNumber;
    }

    public int getDuration() {
        return duration;
    }

    public int getYear() {
        return year;
    }

    public boolean isDirectory() {
        return isDirectory;
    }

    public static class Directory extends CollectionItem {
        public static class Deserializer implements JsonDeserializer<CollectionItem> {
            public Directory deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
                Directory item = new Directory();
                item.isDirectory = true;
                JsonObject fields = json.getAsJsonObject();
                item.parseFields(fields);
                return item;
            }
        }
    }

    public static class Song extends CollectionItem {
        public static class Deserializer implements JsonDeserializer<CollectionItem> {
            public Song deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
                Song item = new Song();
                item.isDirectory = false;
                JsonObject fields = json.getAsJsonObject();
                item.parseFields(fields);
                return item;
            }
        }
    }

    public static class Deserializer implements JsonDeserializer<CollectionItem> {
        public CollectionItem deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            CollectionItem item = new CollectionItem();
            item.isDirectory = json.getAsJsonObject().has("Directory");
            JsonObject fields;
            if (item.isDirectory()) {
                fields = json.getAsJsonObject().get("Directory").getAsJsonObject();
            } else {
                fields = json.getAsJsonObject().get("Song").getAsJsonObject();
            }
            item.parseFields(fields);
            return item;
        }
    }
}

File: app/src/main/java/agersant/polaris/CollectionItem.kt
package agersant.polaris

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.CompositeDecoder
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.encoding.decodeStructure
import kotlinx.serialization.encoding.encodeStructure

@Serializable(with = CollectionItem.Serializer::class)
sealed class CollectionItem {
    abstract val path: String
    abstract val artist: String?
    abstract val artwork: String?
    abstract val album: String?
    abstract val year: Int

    val isDirectory: Boolean
        get() = this is Directory

    val name: String
        get() {
            val chunks = path.split(Regex("[/\\\\]")).filter(String::isNotEmpty)
            return chunks.last()
        }

    object Serializer : KSerializer<CollectionItem> {
        override val descriptor: SerialDescriptor
            get() = buildClassSerialDescriptor("CollectionItem") {
                element<Song>("Song", isOptional = true)
                element<Directory>("Directory", isOptional = true)
            }

        override fun deserialize(decoder: Decoder): CollectionItem = decoder.decodeStructure(descriptor) {
            while (true) {
                when (val index = decodeElementIndex(descriptor)) {
                    0 -> return decodeSerializableElement(descriptor, index, Song.serializer())
                    1 -> return decodeSerializableElement(descriptor, index, Directory.serializer())
                    CompositeDecoder.DECODE_DONE -> break
                }
            }
            error("Missing field 'Song' or 'Directory'")
        }

        override fun serialize(encoder: Encoder, value: CollectionItem) = encoder.encodeStructure(descriptor) {
            when (value) {
                is Song -> encodeSerializableElement(descriptor, 0, Song.serializer(), value)
                is Directory -> encodeSerializableElement(descriptor, 1, Directory.serializer(), value)
            }
        }
    }
}

@Serializable
data class Song(
    @SerialName("path") override val path: String,
    @SerialName("artist") override val artist: String? = null,
    @SerialName("artwork") override val artwork: String? = null,
    @SerialName("album") override val album: String? = null,
    @SerialName("year") override val year: Int = -1,
    @SerialName("title") val title: String? = null,
    @SerialName("album_artist") val albumArtist: String? = null,
    @SerialName("composer") val composer: String? = null,
    @SerialName("lyricist") val lyricist: String? = null,
    @SerialName("genre") val genre: String? = null,
    @SerialName("label") val label: String? = null,
    @SerialName("track_number") val trackNumber: Int = -1,
    @SerialName("disc_number") val discNumber: Int = -1,
    @SerialName("duration") val duration: Int = -1,
) : CollectionItem()

@Serializable
data class Directory(
    @SerialName("path") override val path: String,
    @SerialName("artist") override val artist: String? = null,
    @SerialName("artwork") override val artwork: String? = null,
    @SerialName("album") override val album: String? = null,
    @SerialName("year") override val year: Int = -1,
    @SerialName("date_added") val dateAdded: Int = -1,
) : CollectionItem()
